### DISCUSSION TITLE
how to bind the Dark Mode with other element?

### DISCUSSION THREAD

**Discussion by MohAid** [zauberzeug/nicegui#2004] [Q&A] [answered]
### Question

Hi, could I get an idea about how to enable and disable the Dark Mode for elements (like "header") and binding with `ui.dark_mode().value`?
And how to set this element color for both the enabled and the disabled Dark Mode?

**Comment by phoskee** [comment]
```py
from nicegui import ui

dark_mode = ui.dark_mode(False)
ui.header().classes('dark:bg-yellow-200')
ui.switch('Dark mode').bind_value(dark_mode)

ui.run()

```

you can set the default color and then with `dark:` what you would like with dark-mode.
Keep in mind that not all colors seem to work, and you can find more info here  (#1830)

**Reply by MohAid** [reply]
Thank you @phoskee, I started as you advised ...
- it was hard to control the colors in both Dark mode and Light mode using classes so I used style.
- I used a button to switch and I faced an issue when switching its icon, do you have an idea how to solve that please?
- and through working on that I wondered if it is possible to bind the two string values of the style to the boolean value that is saved in the storage.user dictionary (especially that ui.dark_mode is bindable to it !)?
- and generally any suggestion you give to me to let the code be more niceguic will be appreciated. 

Here is the code:

```py
from nicegui import ui, app

@ui.page('/')
def main_page() -> None:
    header_light_color = 'background-color: #FFFFFF'
    header_dark_color = 'background-color: #121212'  # bg-neutral-950
    header_color = header_light_color
    if app.storage.user.get('dark_mode', False):
        header_color = header_dark_color
        ui.dark_mode().enable()

    with ui.header(elevated=True).style(header_color) as header:
        dark_btn = ui.button(on_click=lambda: turn_dark_mode(), icon='dark_mode')\
            .props('outline round').tooltip('Dark Mode')

    def turn_dark_mode():
        if app.storage.user.get('dark_mode'):
            app.storage.user['dark_mode'] = False
            ui.dark_mode().disable()
            header.style(header_light_color)
            dark_btn.icon = 'light_mode'  # Todo: this is not working !
        else:
            app.storage.user['dark_mode'] = True
            ui.dark_mode().enable()
            header.style(header_dark_color)
            dark_btn.icon = 'dark_mode'  # Todo: this is not working !

ui.run(host="localhost", storage_secret='1111111111111111111', dark=False)
```

**Reply by phoskee** [reply]
>* I used a button to switch and I faced an issue when switching its icon, do you have an idea how to solve that please?

```py
from nicegui import ui, app


@ui.page('/')
def main_page() -> None:

    header_light_color = 'background-color: #FFFFFF'
    header_dark_color = 'background-color: #121212'  # bg-neutral-950
    header_color = header_light_color
    if app.storage.user.get('dark_mode', False):
        header_color = header_dark_color
        ui.dark_mode().enable()

    with ui.header(elevated=True).style(header_color) as header:
        dark_btn = ui.button(on_click=lambda: turn_dark_mode(), icon='dark_mode')\
            .props('outline round').tooltip('Dark Mode')
    def turn_dark_mode():
        if app.storage.user.get('dark_mode'):
            app.storage.user['dark_mode'] = False
            ui.dark_mode().disable()
            header.style(header_light_color)
            dark_btn.props('icon="light_mode"')  # Todo: DONE !
        else:
            app.storage.user['dark_mode'] = True
            ui.dark_mode().enable()
            header.style(header_dark_color) 
            dark_btn.props('icon="dark_mode"')   # Todo: DONE !


ui.run(host="localhost", storage_secret='1111111111111111111', dark=False)
```

**Reply by falkoschindler** [reply, marked as answer]
I would implement it like this:
```py
@ui.page('/')
def page():
    dark_mode = ui.dark_mode().bind_value(app.storage.user, 'dark_mode')
    with ui.header(elevated=True).classes('bg-[#ffffff] dark:bg-[#121212]'):
        ui.button(icon='dark_mode', on_click=dark_mode.enable) \
            .props('outline round').tooltip('Dark Mode').bind_visibility_from(dark_mode, 'value', value=False)
        ui.button(icon='light_mode', on_click=dark_mode.disable) \
            .props('outline round').tooltip('Light Mode').bind_visibility_from(dark_mode, 'value', value=True)
```

- Binding to `app.storage.user` takes care of reading and writing the mode to storage.
- Tailwind classes like `bg-[#ffffff]` allow specifying arbitrary colors.
- The `dark:` suffix automatically applies classes to dark mode only. This way we don't need to update styling when the mode changes.
- I'd simply use two buttons which show and hide depending on the mode. It might cause some duplication (e.g. for `props`), but avoids the need for change handlers and lets you add another button ("auto" mode) more easily.

This approach is similar to the one on the NiceGUI website:
https://github.com/zauberzeug/nicegui/blob/287455511e079909bf9046068f8d9835d623fe87/main.py#L153-L158

**Reply by MohAid** [reply]
Thanks @phoskee I loved knowing this `.props('icon="dark_mode"')`, although @falkoschindler 's code had covered all I needed and more but yours is also a nice thing I had to know. thanks a lot.

And thank you Dr. @falkoschindler that was wonderful, the most important that I learned to use the source code of "nicegui" as a broad example, the thing I don't know why I missed!
